cppimport/checksum.py

```diff
@@ -1,55 +1,48 @@
-import os
-import pickle
+import json
 import hashlib
-import traceback
+import struct
 
 import cppimport.find
 import cppimport.config
 from cppimport.filepaths import make_absolute
 
-# I use .${filename}.cppimporthash as the checksum file for each module.
-def get_checksum_filepath(filepath):
-    return os.path.join(
-        os.path.dirname(filepath),
-        '.' + os.path.basename(filepath) + '.cppimporthash'
-    )
+
+_TAG = b'cppimport'
+_FMT = struct.Struct('q' + str(len(_TAG)) + 's')
 
 def calc_cur_checksum(file_lst, module_data):
     text = b""
     for filepath in file_lst:
-    	with open(filepath, 'rb') as f:
+        with open(filepath, 'rb') as f:
             text += f.read()
     return hashlib.md5(text).hexdigest()
 
 # Use a checksum to see if the file has been changed since the last compilation
 def is_checksum_current(module_data):
     try:
-        checksum_filepath = get_checksum_filepath(module_data['filepath'])
-
-        if not os.path.exists(checksum_filepath):
-            return False
-
+        with open(module_data['ext_path'], 'rb') as f:
+            f.seek(-_FMT.size, 2)
+            json_len, tag = _FMT.unpack(f.read(_FMT.size))
+            if tag != _TAG:
+                cppimport.config.quiet_print("Missing trailer tag")
+                return False
+            f.seek(-(_FMT.size + json_len), 2)
+            json_s = f.read(json_len)
         try:
-            with open(checksum_filepath, 'rb') as f:
-                deps, old_checksum = pickle.load(f)
-        except (ValueError, pickle.UnpicklingError) as e:
-            cppimport.config.quiet_print(
-                "Failed to load checksum due to exception" + traceback.format_exc()
-            )
+            deps, old_checksum = json.loads(json_s)
+        except ValueError:
+            cppimport.config.quiet_print("Failed to load trailer info")
```

Comment:
Can we make this a bit more informative? Perhaps, "Failed to load checksum trailer info from already existing compiled extensions. Rebuilding the extension."

Reply:
done